packages/vite/src/module-runner/moduleCache.ts

```diff
@@ -1,139 +1,128 @@
-    return this.setByModuleId(this.normalize(fsPath), mod)
+  /**
+   * Returns the module node by the URL that was used in the import statement.
+   * Unlike module graph on the server, the URL is not resolved and is used as is.
+   * @param url Server URL that was used in the import statement
+   */
+  public getModuleByUrl(url: string): ModuleRunnerNode | undefined {
+    return this.urlToIdModuleMap.get(unwrapId(url))
   }
 
-  getByModuleId(modulePath: string): ModuleCache {
-    if (!super.has(modulePath)) this.setByModuleId(modulePath, {})
-
-    const mod = super.get(modulePath)!
-    if (!mod.imports) {
-      Object.assign(mod, {
-        imports: new Set(),
-        importers: new Set(),
-        timestamp: 0,
-      })
+  /**
+   * Ensure that module is in the graph. If the module is already in the graph,
+   * it will return the existing module node. Otherwise, it will create a new
+   * module node and add it to the graph.
+   * @param id Resolved module ID
+   * @param url URL that was used in the import statement
+   */
+  public ensureModule(id: string, url: string): ModuleRunnerNode {
+    id = normalizeModuleId(id)
+    if (this.idToModuleMap.has(id)) {
+      const moduleNode = this.idToModuleMap.get(id)!
+      this.urlToIdModuleMap.set(url, moduleNode)
+      return moduleNode
     }
-    return mod
-  }
-
-  override get(fsPath: string): ModuleCache {
-    return this.getByModuleId(this.normalize(fsPath))
-  }
-
-  deleteByModuleId(modulePath: string): boolean {
-    return super.delete(modulePath)
-  }
-
-  override delete(fsPath: string): boolean {
-    return this.deleteByModuleId(this.normalize(fsPath))
-  }
-
-  invalidateUrl(id: string): void {
-    const module = this.get(id)
-    this.invalidateModule(module)
+    const moduleNode = new ModuleRunnerNode(id, url)
+    this.idToModuleMap.set(id, moduleNode)
+    this.urlToIdModuleMap.set(url, moduleNode)
+
+    const fileModules = this.fileToModulesMap.get(moduleNode.file) || new Set()
+    fileModules.add(moduleNode)
+    this.fileToModulesMap.set(moduleNode.file, fileModules)
+    return moduleNode
   }
 
-  invalidateModule(module: ModuleCache): void {
-    module.evaluated = false
-    module.meta = undefined
-    module.map = undefined
-    module.promise = undefined
-    module.exports = undefined
+  public invalidateModule(node: ModuleRunnerNode): void {
+    node.evaluated = false
+    node.meta = undefined
+    node.map = undefined
+    node.promise = undefined
+    node.exports = undefined
     // remove imports in case they are changed,
     // don't remove the importers because otherwise it will be empty after evaluation
     // this can create a bug when file was removed but it still triggers full-reload
     // we are fine with the bug for now because it's not a common case
-    module.imports?.clear()
-  }
-
-  /**
-   * Invalidate modules that dependent on the given modules, up to the main entry
-   */
-  invalidateDepTree(
```

Comment:
I was using `invalidateDepTree` in my example for rsc environment's node runner https://github.com/hi-ogawa/vite-environment-examples/blob/edf04cf23fb472a9d085e1a424ae7543164b98f8/examples/react-server/vite.config.ts#L173 but it doesn't look necessary since it's same as default non-hmr invalidation, so I just removed it https://github.com/hi-ogawa/vite-environment-examples/pull/109